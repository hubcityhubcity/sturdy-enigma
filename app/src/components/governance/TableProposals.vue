<template>
  <div>
    <table class="data-table">
      <tbody>
        <LiProposal
          v-for="proposal in proposals"
          :key="proposal.id"
          :proposal="proposal"
        />
      </tbody>
    </table>
  </div>
</template>

<script>
import LiProposal from "governance/LiProposal"
export default {
  name: `table-proposals`,
  components: {
    LiProposal,
  },
  props: {
    proposals: {
      type: Array,
      required: true,
    },
  },
  data: () => ({
    sort: {
      property: `id`,
      order: `desc`,
    },
  }),
}
</script>
