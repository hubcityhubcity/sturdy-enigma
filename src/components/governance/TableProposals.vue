<template>
  <div>
    <table class="data-table">
      <tbody>
        <LiProposal
          v-for="(value, key) in filteredProposals"
          :key="key"
          :proposal="value"
        />
      </tbody>
    </table>
  </div>
</template>

<script>
import orderBy from "lodash.orderby"
import LiProposal from "governance/LiProposal"
export default {
  name: `table-proposals`,
  components: {
    LiProposal
  },
  props: {
    proposals: {
      type: Object,
      required: true
    }
  },
  data: () => ({
    sort: {
      property: `id`,
      order: `desc`
    }
  }),
  computed: {
    enrichedProposals() {
      const copy = JSON.parse(JSON.stringify(this.proposals))
      Object.keys(copy).forEach(proposal_id => {
        copy[proposal_id].id = Number(proposal_id)
      })
      return copy
    },
    filteredProposals() {
      const proposals = orderBy(
        this.enrichedProposals,
        [this.sort.property],
        [this.sort.order]
      )

      return proposals
    }
  },
  mounted() {
    this.$store.dispatch(`getProposals`)
  }
}
</script>
