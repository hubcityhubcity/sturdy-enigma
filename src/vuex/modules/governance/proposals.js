import Vue from "vue"
import BigNumber from "bignumber.js"

export const setProposalTally = (commit, node) => async proposal => {
  commit(`setProposal`, proposal)
  const final_tally_result =
    proposal.proposal_status === `VotingPeriod`
      ? await node.get.proposalTally(proposal.proposal_id)
      : { ...proposal.final_tally_result }
  commit(`setProposalTally`, {
    proposal_id: proposal.proposal_id,
    final_tally_result
  })
}

export default ({ node }) => {
  const emptyState = {
    loading: false,
    loaded: false,
    error: null,
    proposals: {},
    tallies: {}
  }
  const state = JSON.parse(JSON.stringify(emptyState))
  const mutations = {
    setProposal(state, proposal) {
      Vue.set(state.proposals, proposal.proposal_id, proposal)
    },
    setProposalTally(state, { proposal_id, final_tally_result }) {
      Vue.set(state.tallies, proposal_id, final_tally_result)
    }
  }
  const actions = {
    async reconnected({ state, dispatch }) {
      if (state.loading) {
        await dispatch(`getProposals`)
      }
    },
    async getProposals({ state, commit, rootState }) {
      state.loading = true
      if (!rootState.connection.connected) return

      try {
        const proposals = await node.get.proposals()
        if (proposals.length > 0) {
          await Promise.all(proposals.map(setProposalTally(commit, node)))
        }

        state.error = null
        state.loaded = true
        state.loading = false
      } catch (error) {
        state.error = error
      }
    },
    async getProposal({ state, commit }, proposal_id) {
      state.loading = true
      try {
        const proposal = await node.get.proposal(proposal_id)
        await setProposalTally(commit, node)(proposal)
        state.error = null
        state.loaded = true
        state.loading = false
        return proposal
      } catch (error) {
        state.error = error
      }
      return undefined
    },
    async postMsgSubmitProposal(
      {
        state,
        rootState: { wallet },
        dispatch,
        commit
      },
      {
        txProps: { initialDeposits, title, description }
      }
    ) {
      // optimistic updates
      initialDeposits.forEach(({ amount, denom }) => {
        const oldBalance = wallet.balances.find(
          balance => balance.denom === denom
        )
        commit(`updateWalletBalance`, {
          denom,
          amount: BigNumber(oldBalance.amount)
            .minus(amount)
            .toNumber()
        })
      })

      const latestId = Object.keys(state.proposals).reduce((latest, id) => {
        return latest > Number(id) ? latest : Number(id)
      }, 0)

      commit(`setProposal`, {
        proposal_id: String(latestId + 1),
        proposal_content: {
          value: { title, description }
        },
        initial_deposit: initialDeposits
      })

      await dispatch(`getProposals`)
      await dispatch(`getAllTxs`)
    }
  }
  return {
    state,
    actions,
    mutations
  }
}
