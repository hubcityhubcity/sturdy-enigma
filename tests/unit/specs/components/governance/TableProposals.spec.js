import { mount, createLocalVue } from "@vue/test-utils"
import TableProposals from "governance/TableProposals"

import { proposals } from "../../store/json/proposals"

describe(`TableProposals`, () => {
  let wrapper, $store
  const localVue = createLocalVue()
  localVue.directive(`tooltip`, () => {})

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn()
    }

    wrapper = mount(TableProposals, {
      localVue,
      propsData: { proposals },
      mocks: {
        $store
      },
      stubs: [`router-link`]
    })
  })

  it(`has the expected html structure`, async () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
