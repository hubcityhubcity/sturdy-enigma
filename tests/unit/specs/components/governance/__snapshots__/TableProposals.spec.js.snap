// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableProposals has the expected html structure 1`] = `
<div>
  <table
    class="data-table"
  >
    <tbody>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      Testing proposal from BD
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      test sending proposal
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/35"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      A new test proposal
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      This is a proposal send via Big Dipper with ledger
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/34"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      test
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      test
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/33"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      title
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      this is the description 
blah blah blah
it's really long
Instrument cultivated alteration any favourable expression law far nor. Both new like tore but year. An from mean on with when sing pain. Oh toâ€¦
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/32"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      test submit proposal
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      hello world
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/31"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      Hello Cosmos
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      Cosmic Girl attempting communication
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/28"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      Hello
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      Cosmic
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/27"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      test 2
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      test
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/24"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      test submit proposal
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      test submit proposal desc
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/23"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      test submit proposal
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      test submit proposal desc
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/22"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      test submit proposal
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      test submit proposal desc
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/21"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      test submit proposal
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      test submit proposal desc
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/20"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      prop test v34
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      test gov proposal after updating to 34.0
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/17"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      kiostest
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      vote test
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/15"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      Test
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      This is a test proposal
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/14"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      test
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      d
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/13"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
      <tr
        class="li-proposal"
      >
        <td>
          <span
            class="proposal-status rejected"
          >
            
      Rejected
    
          </span>
           
          <h3
            class="li-proposal-title"
          >
            
      test 2
    
          </h3>
           
          <p
            class="li-proposal-description"
          >
            
      test 2
    
          </p>
           
          <router-link-stub
            class="read-more-link"
            to="/proposals/12"
          >
            Read the full proposal...
          </router-link-stub>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
